product_multi_image/models/product_product.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+# © 2016 Pedro M. Baeza <pedro.baeza@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl-3).
+
+from openerp import api, fields, models
+from openerp.osv import orm
+from openerp.osv import fields as old_fields
+
+
+class ProductProduct(models.Model):
+    _name = "product.product"
+    _inherit = [_name, "base_multi_image.owner"]
+
+    # Make this field computed for getting only the available images
+    image_ids = fields.One2many(
+        compute="_compute_image_ids", comodel_name="base_multi_image.image",
+        inverse="_inverse_image_ids")
+    image_main = fields.Binary(inverse="_inverse_main_image_large")
+    image_main_medium = fields.Binary(inverse="_inverse_main_image_medium")
+    image_main_small = fields.Binary(inverse="_inverse_main_image_small")
+
+    @api.multi
+    def _inverse_main_image(self, image):
+        for product in self:
+            if image:
+                product.image_ids[0].write({
+                    'file_db_store': image,
+                    'storage': 'db',
+                })
+            else:
+                product.image_ids = [(3, product.image_ids[0].id)]
+
+    @api.multi
+    def _inverse_main_image_large(self):
+        for product in self:
+            product._inverse_main_image(product.image_main)
+
+    @api.multi
+    def _inverse_main_image_medium(self):
+        for product in self:
+            product._inverse_main_image(product.image_main_medium)
+
+    @api.multi
+    def _inverse_main_image_small(self):
+        for product in self:
+            product._inverse_main_image(product.image_main_small)
+
+    @api.multi
+    @api.depends('product_tmpl_id', 'product_tmpl_id.image_ids',
+                 'product_tmpl_id.image_ids.product_variant_ids')
+    def _compute_image_ids(self):
+        for product in self:
+            images = product.product_tmpl_id.image_ids.filtered(
+                lambda x: (not x.product_variant_ids or
+                           product.id in x.product_variant_ids.ids))
+            product.image_ids = [(6, 0, images.ids)]
+
+    @api.multi
+    def _inverse_image_ids(self):
+        for product in self:
+            # Remember the list of images that were before changes
+            previous_images = product.product_tmpl_id.image_ids.filtered(
+                lambda x: (not x.product_variant_ids or
+                           product.id in x.product_variant_ids.ids))
+            for image in product.image_ids:
+                if isinstance(image.id, models.NewId):
+                    # Image added
+                    image.owner_id = product.product_tmpl_id.id
+                    image.owner_model = "product.template"
+                    image.product_variant_ids = [(6, 0, product.ids)]
```

Comment:
In new api AFAIK, you do not need all that. If it's a NewId instance it gets created, otherwise it gets linked. All is automatic.